File: src/helper/adminGuard.ts
import { readFileSync, writeFileSync, existsSync } from "node:fs";
import { resolve } from "node:path";
import type { CommandInteraction } from "discord.js";

interface AllowlistEntry {
  guildId: string;
  userId: string;
  username: string;
}

const ALLOWLIST_PATH = resolve(process.cwd(), "allowlist.txt");

function parseAllowlist(): AllowlistEntry[] {
  if (!existsSync(ALLOWLIST_PATH)) return [];

  const content = readFileSync(ALLOWLIST_PATH, "utf-8");
  const entries: AllowlistEntry[] = [];

  for (const line of content.split("\n")) {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith("//")) continue;

    const match = trimmed.match(/^(\d+):\s*(\d+),\s*(.+)$/);
    if (match) {
      entries.push({
        guildId: match[1],
        userId: match[2],
        username: match[3].trim(),
      });
    }
  }

  return entries;
}

function saveAllowlist(entries: AllowlistEntry[]): void {
  const lines = entries.map(
    (e) => `${e.guildId}: ${e.userId}, ${e.username}`,
  );
  writeFileSync(
    ALLOWLIST_PATH,
    "// allowlist.txt\n// Format: guildId: userId, username\n// One entry per line\n\n" +
      lines.join("\n") +
      "\n",
    "utf-8",
  );
}

export function isUserAdmin(userId: string, guildId: string): boolean {
  if (process.env.BOT_CREATOR === userId) return true;
  const entries = parseAllowlist();
  return entries.some(
    (e) => e.guildId === guildId && e.userId === userId,
  );
}

export function requireAdmin(
  interaction: CommandInteraction,
): boolean {
  if (!interaction.guildId) return false;
  const allowed = isUserAdmin(interaction.user.id, interaction.guildId);
  if (!allowed) {
    void interaction.reply({
      content: "You do not have permission to use this command.",
      ephemeral: true,
    });
  }
  return allowed;
}

export function addToAllowlist(
  guildId: string,
  userId: string,
  username: string,
): boolean {
  const entries = parseAllowlist();
  if (entries.some((e) => e.guildId === guildId && e.userId === userId)) {
    return false;
  }
  entries.push({ guildId, userId, username });
  saveAllowlist(entries);
  return true;
}

File: utils/admin_guard.py
import os
from discord.ext import commands
from discord import app_commands


ALLOWLIST_PATH = "allowlist.txt"


def _parse_allowlist():
    entries = []
    try:
        with open(ALLOWLIST_PATH) as f:
            for line in f:
                line = line.strip()
                if not line or line.startswith("//"):
                    continue
                parts = line.split(":", 1)
                if len(parts) != 2:
                    continue
                guild_id = parts[0].strip()
                rest = parts[1].strip()
                user_parts = rest.split(",", 1)
                if len(user_parts) != 2:
                    continue
                user_id = user_parts[0].strip()
                entries.append((guild_id, user_id))
    except FileNotFoundError:
        pass
    return entries


def is_user_admin(user_id: str, guild_id: str) -> bool:
    if os.environ.get("BOT_CREATOR") == user_id:
        return True
    return (guild_id, user_id) in _parse_allowlist()


def add_to_allowlist(guild_id: str, user_id: str, username: str):
    entries = _parse_allowlist()
    if (guild_id, user_id) in entries:
        return False
    with open(ALLOWLIST_PATH, "a") as f:
        f.write(f"{guild_id}: {user_id}, {username}\n")
    return True


def admin_check():
    async def predicate(interaction: discord.Interaction):
        if not interaction.guild_id:
            await interaction.response.send_message(
                "This command can only be used in a server.", ephemeral=True
            )
            return False
        if not is_user_admin(str(interaction.user.id), str(interaction.guild_id)):
            await interaction.response.send_message(
                "You do not have permission to use this command.", ephemeral=True
            )
            return False
        return True
    return app_commands.check(predicate)
